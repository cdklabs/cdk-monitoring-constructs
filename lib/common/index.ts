export * from "./alarm";
export * from "./metric";
export * from "./monitoring";
export * from "./url";
export * from "./widget";
export * from "./strings";
