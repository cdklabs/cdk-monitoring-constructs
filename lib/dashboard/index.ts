export * from "./BitmapDashboard";
export * from "./DashboardRenderingPreference";
export * from "./DynamicDashboardSegment";
export * from "./DashboardSegment";
export * from "./DashboardWithBitmapCopy";
export * from "./DefaultDashboardFactory";
export * from "./DynamicDashboardFactory";
export * from "./DefaultWidgetFactory";
export * from "./IDashboardFactory";
export * from "./IDynamicDashboardFactory";
export * from "./IWidgetFactory";
export * from "./MonitoringNamingStrategy";
export * from "./SingleWidgetDashboardSegment";
export * from "./widget";
