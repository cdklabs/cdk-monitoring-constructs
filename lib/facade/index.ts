export * from "./IMonitoringAspect";
export * from "./MonitoringFacade";
export * from "./MonitoringAspect";
