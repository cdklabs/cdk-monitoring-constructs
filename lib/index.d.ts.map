{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":"AAAA,cAAc,UAAU,CAAC;AACzB,cAAc,aAAa,CAAC;AAC5B,cAAc,UAAU,CAAC;AACzB,cAAc,cAAc,CAAC"}