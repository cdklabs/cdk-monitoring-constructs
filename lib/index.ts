export * from "./common";
export * from "./dashboard";
export * from "./facade";
export * from "./monitoring";
