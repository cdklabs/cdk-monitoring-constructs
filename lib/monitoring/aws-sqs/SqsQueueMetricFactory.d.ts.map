{"version":3,"file":"SqsQueueMetricFactory.d.ts","sourceRoot":"","sources":["SqsQueueMetricFactory.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,MAAM,EAAE,MAAM,qBAAqB,CAAC;AAE7C,OAAO,EAAE,aAAa,EAAmB,MAAM,cAAc,CAAC;AAE9D,MAAM,WAAW,0BAA0B;IACzC,QAAQ,CAAC,KAAK,EAAE,MAAM,CAAC;CACxB;AAED,qBAAa,qBAAqB;IAChC,SAAS,CAAC,QAAQ,CAAC,aAAa,EAAE,aAAa,CAAC;IAChD,SAAS,CAAC,QAAQ,CAAC,KAAK,EAAE,MAAM,CAAC;gBAErB,aAAa,EAAE,aAAa,EAAE,KAAK,EAAE,0BAA0B;IAK3E,oCAAoC;IAQpC,0BAA0B;IAS1B,yBAAyB;IAOzB,4CAA4C;IAQ5C,+BAA+B;IAQ/B,oBAAoB;IAUpB,qBAAqB;IAYrB,iBAAiB;CAWlB"}