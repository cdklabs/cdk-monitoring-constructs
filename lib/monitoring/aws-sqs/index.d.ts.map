{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":"AAAA,cAAc,yBAAyB,CAAC;AACxC,cAAc,sBAAsB,CAAC;AACrC,cAAc,6BAA6B,CAAC"}