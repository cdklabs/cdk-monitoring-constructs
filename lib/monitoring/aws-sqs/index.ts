export * from "./SqsQueueMetricFactory";
export * from "./SqsQueueMonitoring";
export * from "./SqsQueueMonitoringWithDlq";
