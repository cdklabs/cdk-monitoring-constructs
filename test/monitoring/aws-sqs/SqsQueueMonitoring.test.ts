import { Duration, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Queue } from "aws-cdk-lib/aws-sqs";

import { AlarmWithAnnotation, SqsQueueMonitoring } from "../../../lib";
import { TestMonitoringScope } from "../TestMonitoringScope";

test("snapshot test: no alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const queue = new Queue(stack, "Queue", {
    queueName: "DummyQueue",
  });

  new SqsQueueMonitoring(scope, {
    queue,
  });

  // alternative: use reference

  new SqsQueueMonitoring(scope, {
    queue: Queue.fromQueueAttributes(stack, "DummyQueueRef", {
      queueArn: "arn:aws:sqs:us-east-2:123456789012:DummyQueueRef",
    }),
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: all alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const queue = new Queue(stack, "Queue", {
    queueName: "DummyQueue",
  });

  let numAlarmsCreated = 0;

  new SqsQueueMonitoring(scope, {
    queue,
    addQueueMinSizeAlarm: {
      Warning: {
        minMessageCount: 1,
      },
    },
    addQueueMaxSizeAlarm: {
      Warning: {
        maxMessageCount: 5,
      },
    },
    addQueueMaxMessageAgeAlarm: {
      Warning: {
        maxAgeInSeconds: 100,
      },
    },
    addQueueMaxTimeToDrainMessagesAlarm: {
      Warning: {
        maxTimeToDrain: Duration.hours(6),
      },
    },
    addQueueMinIncomingMessagesAlarm: {
      Warning: {
        minIncomingMessagesCount: 0,
      },
    },
    addQueueMaxIncomingMessagesAlarm: {
      Warning: {
        maxIncomingMessagesCount: 1000,
      },
    },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreated = alarms.length;
      },
    },
  });

  expect(numAlarmsCreated).toStrictEqual(6);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
