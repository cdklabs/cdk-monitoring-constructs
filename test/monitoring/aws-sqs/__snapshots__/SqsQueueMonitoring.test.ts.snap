// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Queue4A7E3555": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "QueueName": "DummyQueue",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ScopeTestDummyQueueQueueIncomingMessagesCountMaxWarningEF044B70": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Number of incoming messages into the queue is too high.",
        "AlarmName": "Test-DummyQueue-Queue-Incoming-Messages-Count-Max-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Incoming",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "NumberOfMessagesSent",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 1000,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyQueueQueueIncomingMessagesCountMinWarningC7CBC2CF": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Number of incoming messages into the queue is too low.",
        "AlarmName": "Test-DummyQueue-Queue-Incoming-Messages-Count-Min-Warning",
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Incoming",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "NumberOfMessagesSent",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyQueueQueueMessageAgeMaxWarning6EBEF7A3": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Age of the oldest message in the queue is too high.",
        "AlarmName": "Test-DummyQueue-Queue-Message-Age-Max-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Age",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "ApproximateAgeOfOldestMessage",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 100,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyQueueQueueMessageCountMaxWarning5CFB392D": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Number of messages in the queue is too high.",
        "AlarmName": "Test-DummyQueue-Queue-Message-Count-Max-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Visible",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "ApproximateNumberOfMessagesVisible",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 5,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyQueueQueueMessageCountMinWarning8CB774D3": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Number of messages in the queue is too low.",
        "AlarmName": "Test-DummyQueue-Queue-Message-Count-Min-Warning",
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Visible",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "ApproximateNumberOfMessagesVisible",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyQueueQueueTimeToDrainMaxWarningD9560D70": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Time to drain messages in the queue is too high.",
        "AlarmName": "Test-DummyQueue-Queue-TimeToDrain-Max-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Expression": "(visibleMessages / (consumptionVolume - incomingVolume)) * (PERIOD(consumptionVolume))",
            "Id": "expr_1",
            "Label": "Time to Drain (seconds)",
          },
          Object {
            "Id": "visibleMessages",
            "Label": "Visible",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "ApproximateNumberOfMessagesVisible",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "incomingVolume",
            "Label": "Incoming",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "NumberOfMessagesSent",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "consumptionVolume",
            "Label": "Deleted",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "QueueName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "Queue4A7E3555",
                        "QueueName",
                      ],
                    },
                  },
                ],
                "MetricName": "NumberOfMessagesDeleted",
                "Namespace": "AWS/SQS",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 21600,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`snapshot test: no alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Queue4A7E3555": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "QueueName": "DummyQueue",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
